File: src/loader/loader.cpp
/* Copyright (c) 2016 Baidu, Inc. All Rights Reserved.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
==============================================================================*/
#include "loader/loader.h"

#include <fstream>
#include <fcntl.h>
#include <zconf.h>
#include <errno.h>
#include <sys/mman.h>

#include "base/matrix.h"

using std::ifstream;
using std::ofstream;

namespace mdl {
    Loader *Loader::_instance = nullptr;

    Loader *Loader::shared_instance() {
        if (_instance == nullptr) {
            _instance = new Loader();
        }
        return _instance;
    }

    char *get_binary_data(string filename) {
        FILE *file = fopen(filename.c_str(), "rb");
        if (file == nullptr) {
            throw_exception("can't open binary file");
        }

        fseek(file, 0, SEEK_END);
        long size = ftell(file);
        if (size <= 0 || size < sizeof(int)) {
            throw_exception("binary file size is too small");
        }
        rewind(file);

        char *data = new char[size];
        size_t bytes_read = fread(data, 1, size, file);
        if (bytes_read != size) {
            throw_exception("read binary file bytes do not match with fseek");
        }
        fclose(file);

        int size_in_header = *(int *) data;
        if (size_in_header != size) {
            throw_exception("binary file size do not match with header meta info");
        }

        return data;
    }

    bool Loader::load(string model_path, string weights_path) {
        _loaded = false;
        if (!_cleared) {
            throw_exception("Loader must be cleared before reloading");
        }
        _cleared = false;
        bool success = load_with_quantification(model_path, weights_path);
        _loaded = success;
        return success;
    }

    unsigned get_string_hash(string str) {
        const unsigned A = 54059;
        const unsigned B = 76963;
        unsigned res = 37;
        for (int i = 0; i < str.size(); i += 10) {
            char ch = str[i];
            res = (res * A) ^ (ch * B);
        }
        return res;
    }

    /**
     * get model string with signature validation
     *
     * @param model_path
     * @return
     */
    string get_model_string_from_file(string model_path) {
        ifstream model_file(model_path);
        if (!model_file) {
            throw_exception("can't open model json file");
        }
        stringstream ss;
        ss << model_file.rdbuf();
        model_file.close();
        string model_file_string = ss.str();
        if (model_file_string.size() <= 0) {
            throw_exception("model file string is empty");
        }
        size_t end_pos = model_file_string.rfind("=");
        if (end_pos == string::npos) {
            throw_exception("can't find end symbol in model file");
        }
        size_t blank_pos = model_file_string.rfind(" ");
        if (blank_pos == string::npos) {
            throw_exception("can't find blank splitter in model file");
        }
        if (blank_pos >= end_pos) {
            throw_exception("blank splitter is behind end symbol in model file");
        }
        string model_string = model_file_string.substr(0, blank_pos);

        // get sign_string from the end of the json file
        string sign_string = model_file_string.substr(blank_pos + 1, end_pos - blank_pos - 1);
        if (model_string.size() <= 0) {
            throw_exception("model string is empty");
        }
        if (sign_string.size() <= 0) {
            throw_exception("model sign string is empty");
        }
        // get the expected_sign_string from model_string
        size_t sign = get_string_hash(model_string);
        ss.str("");
        ss << sign;
        string sign_string_expected = ss.str();
        if (sign_string != sign_string_expected) {
            throw_exception("model sign is not expected");
        }
        return model_string;
    }

    /**
     *
     * @param model_path
     * @param weights_path
     * @return
     */
    /*
     * version: version code
     * model_count: num of layers
     * sizei:num of params for the ith layer
     * min max: the min & max value in the params of each layer
     *
     * ||  int  ||    int     || int   || int  || ...  ||float(2) || float(2) || ... || char(30)   || char(30)  || ... || data
     * ||version|| model_count|| size1 ||size2 || ...  || min max ||  min max || ... || layer_name || layer_name|| ... || data
     */
    bool Loader::load_with_quantification(string model_path, string weights_path) {
        string model_string = get_model_string_from_file(model_path);
        string error;
        _model = Json::parse(model_string, error);
        if (error.size() > 0) {
            throw_exception("json parser error %s", error.c_str());
        }

        for (auto &pair: _model["matrix"].object_items()) {
            string name = pair.first;
            Json config = pair.second;
            Matrix *matrix = new Matrix(config);
            matrix->reallocate();
            if (matrix == nullptr) {
                throw_exception("can't create %s matrix", name.c_str());
            }
            if (name == matrix_name_data) {
                _matrices[matrix_name_test_data] = matrix;
                Matrix *data_matrix = new Matrix(config);
                if (data_matrix == nullptr) {
                    throw_exception("can't create data matrix");
                }
                _matrices[matrix_name_data] = data_matrix;
            } else {
                _matrices[name] = matrix;
            }
        }
        // validate the version code in the json file
        int model_version_from_json = _model["meta"]["model_version"].int_value();
        if (model_version != model_version_from_json) {
            throw_exception("model version from cpp does not match with version from json config");
        }
        char *original_binary_data = get_binary_data(weights_path);
        char *binary_data = original_binary_data;
        if (binary_data == nullptr) {
            throw_exception("can't load binary file");
        }

        try {
            if (model_version != *((int *) binary_data + 1)) {
                throw_exception("model version from cpp does not match with version from binary data");
            }
            int model_count = *((int *) binary_data + 2);
            if (model_count <= 0) {
                throw_exception("model count is too small");
            }
            binary_data = binary_data + 3 * sizeof(int);

            // the num of params of each layer
            vector<int> model_sizes(model_count);
            for (int i = 0; i < model_count; i++) {
                model_sizes[i] = *((int *) binary_data + i);
            }
            binary_data = binary_data + model_count * sizeof(int);
            vector<float> model_mins(model_count);
            vector<float> model_maxs(model_count);
            for (int i = 0; i < model_count; i++) {
                model_mins[i] = *((float *) binary_data + i * 2);
                model_maxs[i] = *((float *) binary_data + i * 2 + 1);
            }
            binary_data = binary_data + model_count * 2 * sizeof(float);
            vector<string> model_names(model_count);
            for (int i = 0; i < model_count; i++) {
                model_names[i] = string((char *) binary_data + i * string_size * sizeof(char));
            }
            binary_data = binary_data + model_count * string_size * sizeof(char);

            uint8_t *uint8_data = (uint8_t *) binary_data;
            for (int i = 0; i < model_count; i++) {
                int model_size = model_sizes[i];
                string model_name = model_names[i];
                if (model_name == matrix_name_data) {
                    model_name = matrix_name_test_data;
                }
                if (_matrices.find(model_name) == _matrices.end()) {
                    throw_exception("can't find %s in matrices when load binary file", model_name.c_str());
                }
                Matrix *matrix = _matrices[model_name];
                if (matrix->count() != model_size) {
                    throw_exception("matrix count does not match between json and binary file");
                }
                // inverse quantification
                const float min_value = model_mins[i];
                const float max_value = model_maxs[i];
                const float factor = (max_value - min_value) / 255.0;
                float *matrix_data = matrix->get_data();
                for (int j = 0; j < model_size / 4; j++) {
                    matrix_data[0] = uint8_data[0] * factor + min_value;
                    matrix_data[1] = uint8_data[1] * factor + min_value;
                    matrix_data[2] = uint8_data[2] * factor + min_value;
                    matrix_data[3] = uint8_data[3] * factor + min_value;
                    matrix_data += 4;
                    uint8_data += 4;
                }
                for (int j = 0; j < model_size % 4; j++) {
                    matrix_data[j] = uint8_data[j] * factor + min_value;
                }
                uint8_data += (model_size % 4);
            }
        } catch (...) {
            if (original_binary_data != nullptr) {
                delete original_binary_data;
                original_binary_data = nullptr;
            }
            throw;
            return false;
        }

        if (original_binary_data != nullptr) {
            delete original_binary_data;
            original_binary_data = nullptr;
        }
        return true;
    }

#ifdef NEED_DUMP
    /**
     * non quantification load method
     * @param model_path
     * @param weights_path
     * @return
     */
    bool Loader::load_without_quantification(string model_path, string weights_path) {
        string model_string = get_model_string_from_file(model_path);
        string error;
        _model = Json::parse(model_string, error);
        if (error.size() > 0) {
            throw_exception("json parser error %s", error.c_str());
        }

        for (auto &pair: _model["matrix"].object_items()) {
            string name = pair.first;
            Json config = pair.second;
            Matrix *matrix = new Matrix(config);
            matrix->reallocate();
            if (name == matrix_name_data) {
                _matrices[matrix_name_test_data] = matrix;
                _matrices[matrix_name_data] = new Matrix(config);
            } else {
                _matrices[name] = matrix;
            }
        }

        int model_version_from_json = _model["meta"]["model_version"].int_value();
        if (model_version != model_version_from_json) {
            return false;
        }
        char *original_binary_data = get_binary_data(weights_path);
        char *binary_data = original_binary_data;
        if (binary_data == nullptr) {
            return false;
        }

        if (model_version != *((int *)binary_data + 1)) {
            return false;
        }
        int model_count = *((int *)binary_data + 2);
        if (model_count <= 0) {
            return false;
        }

        binary_data = binary_data + 3 * sizeof(int);
        vector<int> model_sizes(model_count);
        for (int i = 0; i < model_count; i++) {
            model_sizes[i] = *((int *)binary_data + i);
        }
        binary_data = binary_data + model_count * sizeof(int);
        vector<string> model_names(model_count);
        for (int i = 0; i < model_count; i++) {
            model_names[i] = string((char *)binary_data + i * string_size * sizeof(char));
        }
        binary_data = binary_data + model_count * string_size * sizeof(char);
        float *float_data = (float *)binary_data;
        for (int i = 0; i < model_count; i++) {
            int model_size = model_sizes[i];
            string model_name = model_names[i];
            if (model_name == matrix_name_data) {
                model_name = matrix_name_test_data;
            }
            if (_matrices.find(model_name) == _matrices.end()) {
                return false;
            }
            Matrix *matrix = _matrices[model_name];
            std::copy(float_data, float_data + model_size, matrix->get_data());
            float_data += model_size;
        }

        delete original_binary_data;
        return true;
    }
#endif

    void Loader::clear() {
        _loaded = false;
        for (auto &pair: _matrices) {
            Matrix *matrix = pair.second;
            delete matrix;
        }
        _matrices.clear();
        _cleared = true;
    }
};

File: iOS/MDL/MDL/GPU/Loader.swift
/* Copyright (c) 2017 Baidu, Inc. All Rights Reserved.
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 ==============================================================================*/

import Foundation
import MetalPerformanceShaders

@available(iOS 10.0, *)
class Loader {
    static let share:Loader = Loader()
    private let stringSize: CInt = 30
    var matrices: [String : Matrix] = [:]
    var layerModelDic: [String : LayerModel] = [:]
    
    private init(){}
    
    func clear() {
        matrices.removeAll()
        layerModelDic.removeAll()
    }
    
    func load(device: MTLDevice, modelPath: String, para: (String, Int) -> NetParameterData?) throws -> Model{
        return try loadMatrix(device: device, modelPath: modelPath) { (name, size) -> NetParameterData? in
            return para(name, size)
        }
    }
    
    func load(device: MTLDevice, modelPath: String, weightPath: String) throws -> Model   {
        let model = try loadMatrix(device: device, modelPath: modelPath, para: { (name, size) -> NetParameterData? in
            return  ParameterData(size: size)
        })
        try loadBinary(weightPath: weightPath)
        return model
    }
    
    private func loadMatrix(device: MTLDevice, modelPath: String, para: (String, Int) -> NetParameterData?) throws -> Model{
        let fileManager = FileManager.default
        let data = fileManager.contents(atPath: modelPath)
        
        guard let d = data else {
            throw NetError.loaderError(message: "has no data in the filePath: " + modelPath)
        }
        var jsonStr = try String.init(contentsOfFile: modelPath, encoding: .utf8)
        if jsonStr.contains("="){
            var index = jsonStr.index(before: jsonStr.endIndex)
            while jsonStr[index] != Character.init("}"){
                index = jsonStr.index(before: index)
            }
            jsonStr.removeSubrange(jsonStr.index(after: index)..<jsonStr.endIndex)
        }
        
        guard let jsonData = jsonStr.data(using: .utf8) else {
            throw NetError.loaderError(message: "json string -> json data error, json structure error")
        }
        
        let dic = try JSONSerialization.jsonObject(with: jsonData, options: .allowFragments)
        let model = try convert(originModel: Model(values: dic as! NSDictionary))
        
        func createMatrix(name: String) throws -> Matrix{
            guard let matrixConfig = model.matrix[name] else {
                throw NetError.loaderError(message: "matrix in json lack " + name)
            }
            let matrix = Matrix(device: device, name: name, config: matrixConfig)
            return matrix
        }
        for (index, layer) in model.layer.enumerated(){
            if index == 0 {
                for input in layer.input{
                    let matrix = try createMatrix(name: input)
                    layer.inputMatrices.append(matrix)
                }
            }else{
                if layer.type == LayerModel.concatType {
                    var concatConfig = [0, 0, 0, 0]
                    for input in layer.input {
                        let config = model.matrix[input]
                        concatConfig[0] = config?[0] ?? 0
                        concatConfig[1] += config?[1] ?? 0
                        concatConfig[2] = config?[2] ?? 0
                        concatConfig[3] = config?[3] ?? 0
                    }
                    let matrix = Matrix(device: device, name: layer.name, config: concatConfig)
                    model.matrixDesDic[matrix] = matrix.createImageDes()
                    layer.inputMatrices = [matrix]
                    layer.outputMatrices = [matrix]
                    matrices[layer.output[0]] = matrix
                    
                    
                    var channels: Int = 0
                    for input in layer.input {
                        let previousLayerModel = layerModelDic[input]
                        previousLayerModel?.destinationChannelOffset = channels

                        let previousMatrix = matrices[input]
                        if let inPreMatrix = previousMatrix{
                            model.matrixDesDic[inPreMatrix] = nil
                        }
                        
                        channels += previousMatrix?.channels ?? 0
                        previousLayerModel?.outputMatrices.first?.concatMatrix = matrix
                    }
                    
                }else{
                    for input in layer.input {
                        guard let inputMatrix = matrices[input] else {
                            throw NetError.loaderError(message: "no input matrix in " + layer.name)
                        }
                        
                        layer.inputMatrices.append(inputMatrix)
                    }
                }
            }
            if layer.type != LayerModel.concatType{
                for output in layer.output{
                    let matrix = try createMatrix(name: output)
                    matrices[output] = matrix
                    layerModelDic[output] = layer
                    layer.outputMatrices.append(matrix)
                    model.matrixDesDic[matrix] = matrix.createImageDes()
                }
                for weigth in layer.weight{
                    let matrix = try createMatrix(name: weigth)
                    matrices[weigth] = matrix
                    matrix.data = para(matrix.name, matrix.count())
                }
            }
        }
        return model
    }

    private func loadBinary(weightPath: String) throws{
        let dataGetter = try FileDataGetter<CInt>(filePath: weightPath)
        var binaryData = dataGetter.data
        let modelVersion = binaryData[1]
        let modelCount = binaryData[2]
        
        print("\(modelVersion) --- \(modelCount)")
        binaryData = binaryData.advanced(by: 3)
        
        var modelSizes: [CInt] = []
        for i in 0..<modelCount {
            modelSizes.append(binaryData[i])
        }
        binaryData = binaryData.advanced(by: Int(modelCount))
        
        var floatBinaryData:UnsafeMutablePointer<CFloat>  = unsafeBitCast(binaryData, to: UnsafeMutablePointer<CFloat>.self)
        var modelMins: [CFloat] = []
        var modelMaxs: [CFloat] = []
        for i in 0..<modelCount {
            modelMins.append(floatBinaryData[i * 2])
            modelMaxs.append(floatBinaryData[i * 2 + 1])
        }
        floatBinaryData = floatBinaryData.advanced(by: Int(modelCount) * 2)
        
        var charBinaryData: UnsafeMutablePointer<CChar> = unsafeBitCast(floatBinaryData, to: UnsafeMutablePointer<CChar>.self)
        var tempCharBinaryData = charBinaryData
        
        var modelNames: [String] = []
        for _ in 0..<modelCount {
            let modelName = String.init(cString: tempCharBinaryData)
            modelNames.append(modelName)
            tempCharBinaryData = tempCharBinaryData.advanced(by: Int(stringSize))
        }
        charBinaryData = charBinaryData.advanced(by: Int(modelCount * stringSize))
        
        var uInt8Data: UnsafeMutablePointer<UInt8> = unsafeBitCast(charBinaryData, to: UnsafeMutablePointer<UInt8>.self)
        
        for i in 0..<modelCount {
            let modelSize = modelSizes[i]
            let modelName = modelNames[i]
            
            if modelName == "data"{
                continue
            }
            
            guard let matrix = matrices[modelName] else {
                throw NetError.loaderError(message: "can't find " + modelName + " in matrixs when load binary file")
            }
            
            guard matrix.count() == Int(modelSize) else {
                throw NetError.loaderError(message: "matrix count does not match between json and binary file")
            }
            let minvalue = modelMins[i]
            
            let maxValue = modelMaxs[i]
            let factor: CFloat = (maxValue - minvalue)/255.0
            guard var matrixData = matrix.data?.pointer else{
                throw NetError.loaderError(message: "matrix: " + modelName + " data pointer hasn't initialized")
            }
            
            for _ in 0..<modelSize/4{
                matrixData[0] = CFloat(uInt8Data[0]) * factor + minvalue
                matrixData[1] = CFloat(uInt8Data[1]) * factor + minvalue
                matrixData[2] = CFloat(uInt8Data[2]) * factor + minvalue
                matrixData[3] = CFloat(uInt8Data[3]) * factor + minvalue
                matrixData = matrixData.advanced(by: 4)
                uInt8Data = uInt8Data.advanced(by: 4)
            }
            for j in 0..<modelSize%4{
                matrixData[j] = CFloat(uInt8Data[j]) * factor + minvalue
            }
            uInt8Data = uInt8Data.advanced(by: Int(modelSize%4))
        }
    }
    
    private func convert(originModel: Model) throws -> Model{
        let resultModel = Model(values: [:])
        resultModel.matrix = originModel.matrix
        resultModel.layer = []
        var layerDic: [String : LayerModel] = [:]
        for layer in originModel.layer {
            if let output = layer.output.first, layer.type != LayerModel.reluType{
                layerDic[output] = layer
            }
            if layer.type == LayerModel.reluType {
                guard let input = layer.input.first else {
                    throw NetError.modelDataError(message:"input of " + layer.name + " has no element")
                }
                guard let inputLayer = layerDic[input] else {
                    throw NetError.modelDataError(message: "")
                }
                inputLayer.relu = true
                
            }else{
                resultModel.layer.append(layer)
            }
        }
        return resultModel
    }
}
